### DISCUSSION TITLE
Update Expiration Date at every request

### DISCUSSION THREAD

**Discussion by fstephany** [maxcountryman/tower-sessions#47] [Q&A] [answered]
Is there a way to update the expiration date (of both the cookie and the server-managed session) at every request? Something along the line what is described there: https://stackoverflow.com/questions/35106336/sessions-should-a-session-cookie-be-refreshed-for-each-request

I like "By sending it every time, the app ensures that the session expires two hours after their last page hit.". But is this even desirable security-wise?

**Comment by maxcountryman** [comment, marked as answer]
It's technically possible to do this manually with [`set_expiration_time`](https://docs.rs/tower-sessions/latest/tower_sessions/session/struct.Session.html#method.set_expiration_time). You could automate this with a custom extractor, for example and then use it [as a middleware](https://docs.rs/axum/latest/axum/middleware/fn.from_extractor.html). That said, we could also support this more directly through additional configuration options. From a security perspective, it's dependent on your use case.

**Reply by fstephany** [reply]
Ah cool, i might give it a go at some point. 
But I'll check what other frameworks (eg. Django, rails, laravel) do by default first.

**Reply by maxcountryman** [reply]
This crate is modeled fairly closely to Django's session contrib package, so it should behave similarly (if you find differently, please let me know).

**Reply by fstephany** [reply]
Ok! Thanks for the pointers :+1:

**Reply by fstephany** [reply]
You are right, it seems that by default Django session does _not_ resend the cookie with an updated expiration date. You must activate the feature manually by setting `SESSION_SAVE_EVERY_REQUEST` to get this behavior: https://docs.djangoproject.com/en/4.2/topics/http/sessions/#when-sessions-are-saved

**Reply by maxcountryman** [reply]
I'm open to adding this via a configuration option. So if someone would like to take a crack at that I'm happy to review a PR.

**Reply by fstephany** [reply]
I'm not familiar with tower/services/middleware/axum extractor. I'll give it a try as this looks like the perfect issue to get started.

**Reply by 0rzech** [reply]
> I'm open to adding this via a configuration option. So if someone would like to take a crack at that I'm happy to review a PR.

@maxcountryman What do you think about this: https://github.com/maxcountryman/tower-sessions/compare/main...0rzech:tower-sessions:save-session-on-read ? I did it with new `Expiry` variant, because it looks like saving on every read wouldn't make much sense for existing `Expiry` variants. If this is fine, then I can create a PR after adding some tests to cover new option.

**Reply by maxcountryman** [reply]
I'm not at the computer so it's not easy for me to look at the details. However, changing `Expiry` doesn't sound ideal. 

Shouldn't session persistence (with whatever the backend is) be independent of the expiry?

There's a separate issue around the expiry not being properly associated with the backend (it's not part of `Record`) so while it's entirely unrelated, if we were to make breaking changes then I think we should also consider if that's the time to do that.

**Reply by 0rzech** [reply]
You're right and, to be honest, modifying `Expiry` did feel off for me too. I have updated the code. The above link will show updated version.

Now there's the `always_save` field in `SessionConfig` with dedicated builder function, which makes unmodified sessions saved. This way the `Expiry` enum remains unchanged.

**Reply by maxcountryman** [reply]
That sounds like a more reasonable approach.

**Reply by 0rzech** [reply]
And that's what you've actually suggested in the code. 😉 If you're fine with the current approach, I will proceed with tests and making a PR.

**Reply by maxcountryman** [reply]
Sounds good. Happy to review a PR.

**Reply by 0rzech** [reply]
For anyone interested: related [option](https://docs.rs/tower-sessions/0.13.0/tower_sessions/service/struct.SessionManagerLayer.html#method.with_always_save) has been released with version `0.13.0`.

**Reply by avsaase** [reply]
Hi, I found this discussion looking for a way to extend the session expiration when I don't change the session during requests. The option to save the session on every request solves the problem but requires writing to the session store on every page load. This also negates the performance benefit of the `CachingSessionStore` since you need to write to the database anyway.

A more lightweight approach would be to save a session if the current session expires within a given time frame. For example, with a one hour expiry and a 30 minute stale duration, the session would not be saved more often than every 30 minutes and still guarantee an active session for 30 minutes after the last activity. 

For a simpler configuration it would be nice if you could set it to save the session if a given time since the last save has passed but `Record` doesn't contain the last update time so this is not directly possible.

Of course, the library user could implement this themselves by storing the necessary data within the session with a custom extractor (or layer?, I haven't tried this yet) but I think it would be nice if the library natively supported this. I have a branch that does this so I can make a PR if there's interest.

**Reply by avsaase** [reply]
In case anyone is interested, I created a [gist](https://gist.github.com/avsaase/bb6c23209222ec2b18b441a22621dea1) with a middleware that updates the session expiration if a given period since the last update has passed.
